### PR TITLE
RavenDB-15497 - Infinite loop when timeout is exceeded in WaitForIndexesAfterSaveChanges

### PR DESCRIPTION
**Pull request by grisha-kotler**
https://issues.hibernatingrhinos.com/issue/RavenDB-15497

### CODE REVIEW

**Review by ppekrol — APPROVED** [review]

**Comment by ppekrol** [comment]
Can you add a test for this?

File: src/Raven.Server/Documents/Handlers/BatchHandler.cs [hunk 1]
```diff
@@ -341,39 +341,39 @@ public static async Task WaitForReplicationAsync(DocumentDatabase database, Time
...
+
+            while (true)
             {
-                while (true)
-                {
-                    var hadStaleIndexes = false;
+                var hadStaleIndexes = false;
 
-                    using (context.OpenReadTransaction())
+                using (contextPool.AllocateOperationContext(out DocumentsOperationContext context))
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Handlers/BatchHandler.cs:351]: How moving the context allocation here fixes the infinite loop? We open the transaction on each iteration - that should be sufficient/

Comment (grisha-kotler) [reply src/Raven.Server/Documents/Handlers/BatchHandler.cs:351]: Two issues:
- Infinite loop as long as the relevant indexes are stale and throwOnTimeout is false, here:
https://github.com/ravendb/ravendb/blob/v4.2/src/Raven.Server/Documents/Handlers/BatchHandler.cs#L365
- If we have a long timeout, the memory allocations continue to grow

Comment (arekpalinski) [reply src/Raven.Server/Documents/Handlers/BatchHandler.cs:351]: So the fix for the infinite loop is not here but it was the first commit - https://github.com/ravendb/ravendb/pull/10991/commits/50e4b7e1ec4914317ab23dd929e5057485ba0cc0. Now it's clear. 

Why memory allocations continue to grow - don't we reuse already allocated memory in a given context?

Comment (arekpalinski) [reply src/Raven.Server/Documents/Handlers/BatchHandler.cs:351]: Okay. This is because of https://issues.hibernatingrhinos.com/issue/RavenDB-13589

**Comment by grisha-kotler** [comment]
https://github.com/ravendb/ravendb/pull/10998

**Pull request merged**